File: hdf-converters/shared/typescript/stix.ts
// Shared STIX 2.1 bundle parse/detect helper — the TypeScript peer of
// shared/go/stix.go (kept at parity, no fork). Consumed by the enrichment pass
// and (in a later phase) the CLI source fingerprint.

export type StixObject = Record<string, unknown>;

export interface StixBundle {
  type: string;
  id?: string;
  objects: StixObject[];
}

/** Parse and validate a STIX 2.1 bundle. Objects are kept raw for lossless carriage. */
export function parseStixBundle(input: string): StixBundle {
  if (!input) throw new Error('stix: empty input');
  let raw: unknown;
  try {
    raw = JSON.parse(input);
  } catch (e) {
    throw new Error(`stix: parsing bundle: ${(e as Error).message}`);
  }
  const obj = (raw ?? {}) as Record<string, unknown>;
  if (obj.type !== 'bundle') {
    throw new Error(`stix: not a STIX bundle (type=${JSON.stringify(obj.type)})`);
  }
  if (!Array.isArray(obj.objects)) {
    throw new Error('stix: bundle has no objects[]');
  }
  return {
    type: 'bundle',
    id: typeof obj.id === 'string' ? obj.id : undefined,
    objects: obj.objects as StixObject[],
  };
}

/** Report whether the input looks like a STIX 2.1 bundle, without throwing. */
export function detectStixBundle(input: string): boolean {
  try {
    const o = JSON.parse(input) as Record<string, unknown>;
    return o?.type === 'bundle' && Array.isArray(o.objects);
  } catch {
    return false;
  }
}

/** A STIX object's id ('' if absent). */
export function stixObjectId(obj: StixObject): string {
  return typeof obj.id === 'string' ? obj.id : '';
}

/**
 * The CVE ids a STIX object cites via external_references (source_name "cve").
 * A STIX vulnerability records its CVE this way, not in a native field
 * (STIX 2.1 §4.19).
 */
export function stixObjectCVEs(obj: StixObject): string[] {
  const refs = obj.external_references;
  if (!Array.isArray(refs)) return [];
  const cves: string[] = [];
  for (const r of refs) {
    const ref = r as Record<string, unknown>;
    if (
      typeof ref.source_name === 'string' &&
      ref.source_name.toLowerCase() === 'cve' &&
      typeof ref.external_id === 'string' &&
      ref.external_id
    ) {
      cves.push(ref.external_id);
    }
  }
  return cves;
}

File: hdf-converters/shared/go/stix.go
package shared

import (
	"encoding/json"
	"fmt"
	"strings"
)

// StixBundle is a parsed STIX 2.1 bundle. Objects are kept as raw maps so the
// enrichment pass can carry each one through losslessly (no field normalization).
type StixBundle struct {
	Type    string
	ID      string
	Objects []map[string]interface{}
}

// ParseStixBundle parses and validates a STIX 2.1 bundle. Shared by the enrich
// pass and (in a later phase) the CLI source fingerprint — one implementation,
// no fork.
func ParseStixBundle(input []byte) (*StixBundle, error) {
	if len(input) == 0 {
		return nil, fmt.Errorf("stix: empty input")
	}
	var raw struct {
		Type    string                   `json:"type"`
		ID      string                   `json:"id"`
		Objects []map[string]interface{} `json:"objects"`
	}
	if err := json.Unmarshal(input, &raw); err != nil {
		return nil, fmt.Errorf("stix: parsing bundle: %w", err)
	}
	if raw.Type != "bundle" {
		return nil, fmt.Errorf("stix: not a STIX bundle (type=%q)", raw.Type)
	}
	if raw.Objects == nil {
		return nil, fmt.Errorf("stix: bundle has no objects[]")
	}
	return &StixBundle{Type: raw.Type, ID: raw.ID, Objects: raw.Objects}, nil
}

// DetectStixBundle reports whether the input looks like a STIX 2.1 bundle
// (`{type:"bundle", objects:[…]}`), without erroring. Used by the CLI source
// fingerprint.
func DetectStixBundle(input []byte) bool {
	var probe struct {
		Type    string            `json:"type"`
		Objects []json.RawMessage `json:"objects"`
	}
	if err := json.Unmarshal(input, &probe); err != nil {
		return false
	}
	return probe.Type == "bundle" && probe.Objects != nil
}

// StixObjectID returns a STIX object's id ("" if absent).
func StixObjectID(obj map[string]interface{}) string {
	id, _ := obj["id"].(string)
	return id
}

// StixObjectCVEs returns the CVE ids a STIX object cites via its
// external_references (source_name "cve"). A STIX vulnerability records its CVE
// this way rather than in a native field (STIX 2.1 §4.19).
func StixObjectCVEs(obj map[string]interface{}) []string {
	refs, ok := obj["external_references"].([]interface{})
	if !ok {
		return nil
	}
	var cves []string
	for _, r := range refs {
		ref, ok := r.(map[string]interface{})
		if !ok {
			continue
		}
		if src, _ := ref["source_name"].(string); strings.EqualFold(src, "cve") {
			if id, _ := ref["external_id"].(string); id != "" {
				cves = append(cves, id)
			}
		}
	}
	return cves
}
